subcommands/targets/offline-update.go

```diff
@@ -0,0 +1,286 @@
+	ok, err := checkIfTargetExists(targetName)
+	subcommands.DieNotNil(err, "Failed to check whether Target exists:")
+	if !ok {
+		subcommands.DieNotNil(fmt.Errorf("The specified Target has not been found; target: %s, tag: %s, production: %v\n", targetName, tag, prod))
+	}
+
+	params, err := getTargetInfo(targetName)
+	subcommands.DieNotNil(err, "Failed to obtain Target's details:")
+
+	fmt.Printf("Downloading offline update content of Target %s to %s\n", targetName, dstDir)
+
+	fmt.Println("Downloading TUF metadata...")
+	err = downloadTufRepo(factory, path.Join(dstDir, "tuf"))
+	subcommands.DieNotNil(err, "Failed to download TUF metadata:")
+
+	fmt.Printf("Downloading an ostree repo from the Target's OE build %d...\n", params.ostreeVersion)
+	err = downloadOstree(params.ostreeVersion, params.hardwareID, dstDir)
+	subcommands.DieNotNil(err, "Failed to download Target's ostree repo:")
+
+	fmt.Printf("Downloading Apps fetched by the `assemble-system-image` run; build number:  %d, tag: %s...\n",
+		params.version, params.buildTag)
+	err = downloadApps(targetName, params.version, params.buildTag, path.Join(dstDir, "apps"))
+	subcommands.DieNotNil(err, "Failed to download Target's Apps:")
+
+	fmt.Println("Successfully downloaded offline update content")
+}
+
+func checkIfTargetExists(targetName string) (bool, error) {
+	data, err := api.TufMetadataGet(factory, "targets.json", tag, prod)
+	if herr := client.AsHttpError(err); herr != nil && herr.Response.StatusCode == 404 {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+
+	targets := client.AtsTufTargets{}
+	err = json.Unmarshal(*data, &targets)
+	if err != nil {
+		return false, err
+	}
+	for tn := range targets.Signed.Targets {
+		if tn == targetName {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func getTargetInfo(targetName string) (*targetInfo, error) {
+	// Getting Target's custom info from the `/ota/factories/<factory>/targets/<target-name>` because:
+	// 1. a target name is unique and represents the same Target across all "tagged" targets set including prod;
+	// 2. only this target version/representation contains an original tag(s)/branch that
+	// the `image-assemble` and apps fetching was performed for (needed for determining where to download Apps from).
+	custom, err := getTargetCustomInfo(targetName)
+	if err != nil {
+		return nil, err
+	}
+
+	info := targetInfo{}
+	info.version, err = strconv.Atoi(custom.Version)
+	if err != nil {
+		return nil, err
+	}
+	info.hardwareID = custom.HardwareIds[0]
+	info.buildTag = custom.Tags[0] // See the assemble.py script in ci-scripts https://github.com/foundriesio/ci-scripts/blob/18b4fb154c37b6ad1bc6e7b7903a540b7a758f5d/assemble.py#L300
+	info.ostreeVersion = info.version
+	if len(custom.OrigUri) > 0 {
+		indx := strings.LastIndexByte(custom.OrigUri, '/')
+		if indx == -1 {
+			fmt.Fprintf(os.Stderr, "Failed to determine Target's OE build version: %s", custom.OrigUri)
+			os.Exit(1)
```

Comment:
Why not `return nil, fmt.Errorf(...)`?

The reason we don't do `fmt.Fprintf` for errors is consistency. The `return err` and `DieNotNil` things were called to give our error messages at least some degree of the consistent look-and-feel (the primary intent not to save code lines as one might think).

Reply:
Good catch, it seems like leftover from the previous version.

Reply:
done.